exercises/npm-registry/src/package.ts

```diff
@@ -1,22 +1,98 @@
 import { RequestHandler } from 'express';
 import got from 'got';
+import { getLatestSemVer, isSemVerFilter } from './semver';
 import { NPMPackage } from './types';
 
-/**
- * Attempts to retrieve package data from the npm registry and return it
- */
+/*
+  This in-memory Map is pretty dangerous:
+    - NPMPackage size & count are unbounded so will very likely consume available ram and crash the container
+    - not abiding by HTTP response cache headers, so updated NPM versions won't make it in here until the container restarts
+    - not shared by containers, so huge waste in resources as each has to cache the same entities
+    - new containers will create extra pressure on external resource (npm here) as the cache is empty, so a scaling event or redeploy means a stampeding herd :o
+
+  But it's also a few lines :)
+
+  Next steps would be an external in-line HTTP cache (eg Varnish). Could cluster it, serve stale responses if there's upstream issues, etc.
+
+  Depending on data mutation needs, cache hit rate, response sizes, etc I'd then consider an external memory cache (eg Redis/Memcache) if needed, then a persistant data store.
+*/
+const npmCache: Map<string, NPMPackage> = new Map<string, NPMPackage>();
+
+// METRIC histogram: req.package.client-time (from the client, be it UI or another backend service for a better overall view of latency)
 export const getPackage: RequestHandler = async function (req, res, next) {
   const { name, version } = req.params;
 
   try {
-    const npmPackage: NPMPackage = await got(
-      `https://registry.npmjs.org/${name}`,
-    ).json();
-
-    const dependencies = npmPackage.versions[version].dependencies;
-
-    return res.status(200).json({ name, version, dependencies });
+    const body = await getPackageDetails({[name]: version});
+    // METRIC histogram: req.package.transitive-dependencies (to track graph depth)
+    // METRIC histogram: req.package.time {status: success}
+    return res.status(200).json(body[0]);
   } catch (error) {
+    // METRIC histogram: req.package.time {status: failure}
+    // LOG: error message w/structured req name/version
     return next(error);
   }
 };
+
+async function getPackageDetails(deps: {[packageName: string]: string}): Promise<any[]> {
+  return Promise.all(Object.keys(deps)
+    .map(async name => {
+      const npmPackage = await fetchNpmPackage(name);
+
+      const requestedVersion = deps[name];
+      const version = determineVersion(requestedVersion, npmPackage);
+      const versionDef = npmPackage.versions[version];
+
+      if (!versionDef) {
+        // METRIC counter: req.package.version-not-found
+        throw new Error(`Version '${version} not found for package ${name}`);
+      }
+
+      const dependencies = versionDef.dependencies;
+      const transitiveDeps = dependencies ? await getPackageDetails(dependencies) : undefined;
```

Comment:
Yes, recursive :(  I had a quick go at converting to a stack based iterative approach but the async & my lack of TS experience was doing my head in a bit!